### PR TITLE
Support Scala 3

### PR DESCRIPTION
**Pull request by lolgab**

### CODE REVIEW

File: build.sbt [hunk 1]
```diff
@@ -1,10 +1,14 @@
-Global / onChangedBuildSource := ReloadOnSourceChanges
+val scala3 = "3.1.0"
+val scala213 = "2.13.4"
+val scala212 = "2.12.13"
```
Comment (sideeffffect) [inline build.sbt:3]: Maybe upgrading to the newest would help? :shrug: 3.1.1, 2.13.8 and 2.12.15

**Comment by lolgab** [comment]
In the [first commit](https://github.com/scala-native/scala-native-loop/pull/39/commits/7621b8e666a46d7ae81fa1799dc04a36b5f0b4c5) I get:
```
[error] java.util.NoSuchElementException: key not found: Local(3)
[error] 	at scala.collection.MapLike.default(MapLike.scala:236)
[error] 	at scala.collection.MapLike.default$(MapLike.scala:235)
[error] 	at scala.collection.AbstractMap.default(Map.scala:65)
[error] 	at scala.collection.MapLike.apply(MapLike.scala:144)
[error] 	at scala.collection.MapLike.apply$(MapLike.scala:143)
[error] 	at scala.collection.AbstractMap.apply(Map.scala:65)
[error] 	at scala.scalanative.interflow.State.loadLocal(State.scala:119)
[error] 	at scala.scalanative.interflow.Eval.eval(Eval.scala:868)
[error] 	at scala.scalanative.interflow.Eval.eval$(Eval.scala:865)
[error] 	at scala.scalanative.interflow.Interflow.eval(Interflow.scala:10)
[error] 	at scala.scalanative.interflow.Eval.eval(Eval.scala:383)
[error] 	at scala.scalanative.interflow.Eval.eval$(Eval.scala:107)
[error] 	at scala.scalanative.interflow.Interflow.eval(Interflow.scala:10)
[error] 	at scala.scalanative.interflow.Eval.run(Eval.scala:30)
[error] 	at scala.scalanative.interflow.Eval.run$(Eval.scala:11)
[error] 	at scala.scalanative.interflow.Interflow.run(Interflow.scala:10)
[error] 	at scala.scalanative.interflow.MergeProcessor.visit(MergeProcessor.scala:376)
[error] 	at scala.scalanative.interflow.MergeProcessor.advance(MergeProcessor.scala:393)
[error] 	at scala.scalanative.interflow.Opt.process(Opt.scala:127)
[error] 	at scala.scalanative.interflow.Opt.process$(Opt.scala:111)
[error] 	at scala.scalanative.interflow.Interflow.process(Interflow.scala:10)
[error] 	at scala.scalanative.interflow.Inline.$anonfun$inline$1(Inline.scala:151)
[error] 	at scala.scalanative.interflow.Log.in(Log.scala:18)
[error] 	at scala.scalanative.interflow.Log.in$(Log.scala:8)
[error] 	at scala.scalanative.interflow.Interflow.in(Interflow.scala:10)
[error] 	at scala.scalanative.interflow.Inline.inline(Inline.scala:139)
[error] 	at scala.scalanative.interflow.Inline.inline$(Inline.scala:134)
[error] 	at scala.scalanative.interflow.Interflow.inline(Interflow.scala:10)
[error] 	at scala.scalanative.interflow.Eval.nonIntrinsic$1(Eval.scala:149)
[error] 	at scala.scalanative.interflow.Eval.eval(Eval.scala:163)
[error] 	at scala.scalanative.interflow.Eval.eval$(Eval.scala:107)
[error] 	at scala.scalanative.interflow.Interflow.eval(Interflow.scala:10)
[error] 	at scala.scalanative.interflow.Eval.run(Eval.scala:30)
[error] 	at scala.scalanative.interflow.Eval.run$(Eval.scala:11)
[error] 	at scala.scalanative.interflow.Interflow.run(Interflow.scala:10)
[error] 	at scala.scalanative.interflow.MergeProcessor.visit(MergeProcessor.scala:376)
[error] 	at scala.scalanative.interflow.MergeProcessor.advance(MergeProcessor.scala:393)
[error] 	at scala.scalanative.interflow.Opt.process(Opt.scala:127)
[error] 	at scala.scalanative.interflow.Opt.process$(Opt.scala:111)
[error] 	at scala.scalanative.interflow.Interflow.process(Interflow.scala:10)
[error] 	at scala.scalanative.interflow.Opt.$anonfun$opt$1(Opt.scala:74)
[error] 	at scala.scalanative.interflow.Log.in(Log.scala:18)
[error] 	at scala.scalanative.interflow.Log.in$(Log.scala:8)
[error] 	at scala.scalanative.interflow.Interflow.in(Interflow.scala:10)
[error] 	at scala.scalanative.interflow.Opt.opt(Opt.scala:22)
[error] 	at scala.scalanative.interflow.Opt.opt$(Opt.scala:22)
[error] 	at scala.scalanative.interflow.Interflow.opt(Interflow.scala:10)
[error] 	at scala.scalanative.interflow.Visit.visitMethod(Visit.scala:124)
[error] 	at scala.scalanative.interflow.Visit.visitMethod$(Visit.scala:117)
[error] 	at scala.scalanative.interflow.Interflow.visitMethod(Interflow.scala:10)
[error] 	at scala.scalanative.interflow.Visit.visit$1(Visit.scala:97)
[error] 	at scala.scalanative.interflow.Visit.$anonfun$visitLoop$1(Visit.scala:111)
[error] 	at scala.scalanative.interflow.Visit.$anonfun$visitLoop$1$adapted(Visit.scala:111)
[error] 	at scala.collection.parallel.mutable.ParArray$ParArrayIterator.foreach(ParArray.scala:145)
[error] 	at scala.collection.parallel.ParIterableLike$Foreach.leaf(ParIterableLike.scala:982)
[error] 	at scala.collection.parallel.Task.$anonfun$tryLeaf$1(Tasks.scala:53)
[error] 	at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:23)
[error] 	at scala.util.control.Breaks$$anon$1.catchBreak(Breaks.scala:67)
[error] 	at scala.collection.parallel.Task.tryLeaf(Tasks.scala:56)
[error] 	at scala.collection.parallel.Task.tryLeaf$(Tasks.scala:50)
[error] 	at scala.collection.parallel.ParIterableLike$Foreach.tryLeaf(ParIterableLike.scala:979)
[error] 	at scala.collection.parallel.AdaptiveWorkStealingTasks$WrappedTask.internal(Tasks.scala:160)
[error] 	at scala.collection.parallel.AdaptiveWorkStealingTasks$WrappedTask.internal$(Tasks.scala:157)
[error] 	at scala.collection.parallel.AdaptiveWorkStealingForkJoinTasks$WrappedTask.internal(Tasks.scala:440)
[error] 	at scala.collection.parallel.AdaptiveWorkStealingTasks$WrappedTask.compute(Tasks.scala:150)
[error] 	at scala.collection.parallel.AdaptiveWorkStealingTasks$WrappedTask.compute$(Tasks.scala:149)
[error] 	at scala.collection.parallel.AdaptiveWorkStealingForkJoinTasks$WrappedTask.compute(Tasks.scala:440)
[error] 	at java.util.concurrent.RecursiveAction.exec(RecursiveAction.java:189)
[error] 	at java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:289)
[error] 	at java.util.concurrent.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1056)
[error] 	at java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1692)
[error] 	at java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:175)
[error] (core / Test / nativeLink) java.util.NoSuchElementException: key not found: Local(3)
[error] Total time: 135 s (02:15), completed Jan 24, 2022 9:08:12 AM
```

Which I couldn't understand why it happens.
@WojciechMazur when you have time do you mind having a look?

**Comment by WojciechMazur** [comment]
@lolgab I've seen some similar problems in the past but typically they're quite hard to debug. In most cases, these kinds of errors are caused by either incorrect merging of the state in the Interflow or when some code is incorrectly structured in NIR.  I'll try to check it out today and would report when I would get some info.

**Comment by lolgab** [comment]
@WojciechMazur I found a small reproduction for the error:
```scala
import scala.scalanative.unsafe._

object Main {
  type Func = CFuncPtr2[Ptr[Byte], Int, Unit]
  val func: Func = (handle: Ptr[Byte], i: Int) => ()
  def main(args: Array[String]): Unit = {}
}
```

It seems it happens when `CFuncPtr`s have both pointers and Ints

Opened an issue here: https://github.com/scala-native/scala-native/issues/2550

File: build.sbt (hunk 1 above)
Comment (lolgab) [reply build.sbt:3]: It was a bug a Scala Native unrelated with the Scala version.

**Pull request merged**